gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java

```diff
@@ -1926,6 +1955,35 @@ public int compare(AssignmentOrder ao1, AssignmentOrder ao2) {
         }
       }
     }
+
+	/**
+	 * Comparator class for sorting Assignments in their categorised ordering
+	 */
+	class CategorizedAssignmentComparator implements Comparator<Assignment> {
+		private Map<String, List<Long>> categoryOrder;
+		private List<String> sortedCategories;
+
+		public CategorizedAssignmentComparator(Map<String, List<Long>> categoryOrder) {
+			this.categoryOrder = categoryOrder;
+			this.sortedCategories = new ArrayList(categoryOrder.keySet());
+			Collections.sort(this.sortedCategories);
+		}
+		
+
+		@Override
+		public int compare(Assignment a1, Assignment a2) {
+			if (a1.getCategoryName() == null) {
+				return 1;
+			} else if (a2.getCategoryName() == null) {
+				return -1;
+			} else if (a1.getCategoryName().equals(a2.getCategoryName())) {
+				List<Long> assignmentsInCategory = categoryOrder.get(a1.getCategoryName());
+				return Integer.compare(assignmentsInCategory.indexOf(a1.getId()), assignmentsInCategory.indexOf(a2.getId()));
+			} else {
+				return Integer.compare(sortedCategories.indexOf(a1), sortedCategories.indexOf(a2));
+			}
+		}
+	}
```

Comment:
Might be able to use apache commons CompareToBuilder, though would need to handle that custom case with the assignments.
Let me know if you want to refactor to use that, otherwise could you just stick some docs in there to explain the sorting strategy and what each bit is for, just for future maintenance?